app/models/user_mentee_application.rb

```diff
@@ -28,9 +29,60 @@
 #
 class UserMenteeApplication < ApplicationRecord
   belongs_to :user
+  has_many :mentee_application_states, dependent: :destroy
 
   validates :available_hours_per_week, numericality: { greater_than: 0, less_than_or_equal_to: 60 }
 
   validates :city, :state, :country, :reason_for_applying, :learned_to_code, :project_experience,
     :available_hours_per_week, presence: true
+
+  after_create :create_initial_application_state
+
+  def current_status
+    current_state.status
+  end
+
+  def current_state
+    mentee_application_states.last
+  end
+
+  def next_status
+    current_status_index = MenteeApplicationState::STATUSES.keys.index(current_status.to_sym)
+    next_status_index = current_status_index.to_i + 1
+
+    MenteeApplicationState::STATUSES.keys.fetch(next_status_index)
+  end
+
+  def promote_application(current_user)
+    next_status = self.next_status
+    mentee_application_states.build(status: next_status, status_changed_by_id: current_user.id).save
+  end
+
+  def reject_application(current_user)
+    mentee_application_states.build(status: :rejected, status_changed_by_id: current_user.id).save
+  end
+
+  def can_promote?
+    current_index = MenteeApplicationState::STATUSES.keys.index(last_status)
+    max_index = MenteeApplicationState::STATUSES.keys.length - 2
+    current_index < max_index && last_status != :rejected
+  end
+
+  def rejected?
+    last_status == :rejected
```

Comment:
This can be delegated to `current_state`. Same with the `accepted?` method. Let me know if you want me to explain more about how this works.
```rb
delegate :rejected?, to: :current_state
```